server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -99,18 +111,47 @@ public static KnnVectorQueryBuilder fromXContent(XContentParser parser) {
     private Integer numCands;
     private final List<QueryBuilder> filterQueries = new ArrayList<>();
     private final Float vectorSimilarity;
+    private final QueryVectorBuilder queryVectorBuilder;
+    private final Supplier<float[]> queryVectorSupplier;
 
     public KnnVectorQueryBuilder(String fieldName, float[] queryVector, Integer numCands, Float vectorSimilarity) {
+        this(fieldName, queryVector, null, null, numCands, vectorSimilarity);
+    }
+
+    public KnnVectorQueryBuilder(String fieldName, QueryVectorBuilder queryVectorBuilder, Integer numCands, Float vectorSimilarity) {
+        this(fieldName, null, queryVectorBuilder, null, numCands, vectorSimilarity);
+    }
+
+    public KnnVectorQueryBuilder(String fieldName, Supplier<float[]> queryVectorSupplier, Integer numCands, Float vectorSimilarity) {
+        this(fieldName, null, null, queryVectorSupplier, numCands, vectorSimilarity);
+    }
+
+    public KnnVectorQueryBuilder(
```

Comment:
This should be a `private` constructor. We don't want anybody else other than this class to set `queryVectorSupplier`